### PR TITLE
Improve logging of GC ref map mismatches

### PR DESCRIPTION
**Pull request by trylek**
As part of investigation of the GC refmap mismatch I'm fixing with

https://github.com/dotnet/runtime/pull/80218

I have improved CoreCLR runtime behavior in the presence of GC refmap mismatches by adding debug-only logic to dump both the GC ref map emitted by Crossgen2 and the one calculated by the native runtime. The dump format matches the way R2RDump outputs GC ref maps next to method import cells in the R2R file.

Thanks

Tomas

/cc @dotnet/crossgen-contrib 

C:\git\runtime>C:\triage\wpf\bin\Debug\net7.0-windows\win-arm64\publish\wpf.exe
GC ref map mismatch detected for method: System.Windows.Media.PathGeometry::GetPathBoundsAsRB
  Runtime GC ref map: I(70) R(78) I(80)
Crossgen2 GC ref map: I(68 70) R(78) I(80)

Assert failure(PID 12596 [0x00003134], Thread: 9568 [0x2560]): false

CORECLR! CheckGCRefMapEqual + 0x1DC (0x00007fff`e66fa72c)
CORECLR! ExternalMethodFixupWorker + 0xA98 (0x00007fff`e681b3c8)
CORECLR! DelayLoad_MethodCall + 0x58 (0x00007fff`e6581a18)
PRESENTATIONCORE! <no symbol> + 0x0 (0x00007fff`e401ecac)
PRESENTATIONCORE! <no symbol> + 0x0 (0x00007fff`e401eaa4)
PRESENTATIONFRAMEWORK! <no symbol> + 0x0 (0x00007fff`dbd34db4)
PRESENTATIONFRAMEWORK! <no symbol> + 0x0 (0x00007fff`dbd33d6c)
PRESENTATIONFRAMEWORK! <no symbol> + 0x0 (0x00007fff`dbc9a85c)
PRESENTATIONCORE! <no symbol> + 0x0 (0x00007fff`e3ecde14)
PRESENTATIONFRAMEWORK! <no symbol> + 0x0 (0x00007fff`dc13ce5c)
    File: C:\git\runtime\src\coreclr\vm\frames.cpp Line: 2177
    Image: C:\triage\wpf\bin\Debug\net7.0-windows\win-arm64\publish\wpf.exe

### CODE REVIEW

**Comment by trylek** [comment]
(apologies for updating the bug description, I have originally copied an earlier version of the CoreCLR runtime dump where I had the tags for Crossgen2 vs. Runtime the other way around)

**Comment by janvorli** [comment]
The Unix builds are failing because of the `wprintf` usage. We have recently removed support for `wprintf` from PAL, see https://github.com/dotnet/runtime/pull/77852.

**Comment by trylek** [comment]
@janvorli - thanks for the heads-up, so does that mean that I should be using <code>printf</code> instead?

**Comment by janvorli** [comment]
> so does that mean that I should be using `printf` instead?

Yes

**Comment by janvorli** [comment]
And also not to use `%S` in `printf`, as on Unix it would expect 32 bit char.

File: src/coreclr/vm/frames.cpp [hunk 1]
```diff
@@ -2074,16 +2145,36 @@ bool CheckGCRefMapEqual(PTR_BYTE pGCRefMap, MethodDesc* pMD, bool isDispatchCell
...
+    while (!decoderNew.AtEnd() || !decoderExisting.AtEnd())
+    {
+        if (decoderNew.AtEnd() != decoderExisting.AtEnd())
+        {
+            invalidGCRefMap = true;
+            break;
+        }
+        if (decoderNew.CurrentPos() != decoderExisting.CurrentPos())
+        {
+            invalidGCRefMap = true;
```
[the remark this answers is not in the document]
Comment (trylek) [reply src/coreclr/vm/frames.cpp]: Thanks Jan for the suggestion, you're right, it's no longer necessary - that was part of my initial implementation where I was dumping the mismatches directly in the validation loop but the outputs weren't sufficiently actionable. I have fixed this and simplified the validation loop in 3rd commit.

**Review by janvorli — APPROVED** [review]
LGTM, thank you!

File: src/coreclr/vm/frames.cpp [hunk 2]
```diff
@@ -2060,6 +2062,75 @@ void FakeGcScanRoots(MetaSig& msig, ArgIterator& argit, MethodDesc * pMD, BYTE *
...
+
+                case GCREFMAP_METHOD_PARAM:
+                    printf("M(");
+                    break;
+
+                case GCREFMAP_TYPE_PARAM:
+                    printf("T(");
+                    break;
+
+                case GCREFMAP_VASIG_COOKIE:
```
Comment (janvorli) [inline src/coreclr/vm/frames.cpp:2108]: This case is missing a `break` which causes the Unix builds to fail.

Comment (trylek) [reply src/coreclr/vm/frames.cpp:2108]: OMG, I don't know how I missed that, fixed in the latest iteration, thanks for the heads-up.

**Comment by AntonLapounov** [comment]
Great improvement.  I have a question regarding dumping: we skip `GCREFMAP_SKIP` tokens when we dump; however, in theory it is possible that the two GCRefMaps differ exactly in one of `GCREFMAP_SKIP` tokens, in which case the difference may not be easily seen from the output.

Nit: this may need a trailing space.
```C++
printf("POP(0x%x)", stackPop);
```

**Comment by trylek** [comment]
Thanks Anton for your feedback. I think that <code>GCREFMAP_SKIP</code> can only become an issue at the end of the GC refmap dump, otherwise it would distort the offset and trigger a mismatch for the subsequent non-skip tokens. From this viewpoint I guess we might want to log trailing <code>GCREFMAP_SKIP</code> entries loosely akin to how Roslyn now reports stray whitespace at line ends. I think you're also right about the POP, I "copied" that bug from R2RDump which behaves the same I believe. I'll put up a PR to fix these two issues, thanks for pointing them out!

**Pull request merged**